Applications: HelmClient DonwloadChart() panic when chart version is empty or not a sermver

### What happened?

When calling `HelmClient.DownloadChart` method with a version that is empty or not a sem version, we got a NPE.

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x5530eeb]

goroutine 131 [running]:
testing.tRunner.func1.2({0x59af780, 0x6fab8a0})
	/Users/vince/go/go1.18.3/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/Users/vince/go/go1.18.3/src/testing/testing.go:1392 +0x39f
panic({0x59af780, 0x6fab8a0})
	/Users/vince/go/go1.18.3/src/runtime/panic.go:838 +0x207
helm.sh/helm/v3/pkg/registry.(*Client).Tags(0x0, {0xc00066a276?, 0xc000543600?})
	/Users/vince/workspace/kubermatic/kubermatic/vendor/helm.sh/helm/v3/pkg/registry/client.go:602 +0x12b
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).getOciURI(0xc000543c30, {0xc00066a270, 0x2e}, {0x5c70993, 0x6}, 0xc000668360)
	/Users/vince/workspace/kubermatic/kubermatic/vendor/helm.sh/helm/v3/pkg/downloader/chart_downloader.go:154 +0x129
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).ResolveChartVersion(0xc000543c30, {0xc00066a270, 0x2e}, {0x5c70993, 0x6})
	/Users/vince/workspace/kubermatic/kubermatic/vendor/helm.sh/helm/v3/pkg/downloader/chart_downloader.go:199 +0x12ff
helm.sh/helm/v3/pkg/downloader.(*ChartDownloader).DownloadTo(0xc000543c30, {0xc00066a270, 0x2e}, {0x5c70993?, 0x0?}, {0xc000774320, 0x4a})
	/Users/vince/workspace/kubermatic/kubermatic/vendor/helm.sh/helm/v3/pkg/downloader/chart_downloader.go:90 +0x5b
k8c.io/kubermatic/v2/pkg/applications/helmclient.HelmClient.DownloadChart({{0x5fa28b0, 0xc00005a120}, {0x5fa3f38, 0xc000555000}, {{0xc0003e5320, 0x5c}, {0xc0003e5380, 0x55}}, {0xc00048b580, 0x2, ...}, ...}, ...)

```

--- a/pkg/applications/helmclient/client.go
+++ b/pkg/applications/helmclient/client.go
@@ -93,18 +93,18 @@ func (a *AuthSettings) newRegistryClient() (*registry.Client, error) {
 	return registry.NewClient(registry.ClientOptCredentialsFile(a.RegistryConfigFile))
 }
 
-// getterOptions return authentication options for Getter.
-func (a AuthSettings) getterOptions() ([]getter.Option, error) {
+// registryClientAndGetterOptions return registry.Client and authentication options for Getter.
+func (a *AuthSettings) registryClientAndGetterOptions() (*registry.Client, []getter.Option, error) {
 	regClient, err := a.newRegistryClient()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	options := []getter.Option{getter.WithRegistryClient(regClient)}
 
 	if a.Username != "" && a.Password != "" {
 		options = append(options, getter.WithBasicAuth(a.Username, a.Password))
 	}
-	return options, nil
+	return regClient, options, nil
 }
 
 // HelmClient is a client that allows interacting with Helm.
@@ -172,7 +172,7 @@ func (h HelmClient) DownloadChart(url string, chartName string, version string, 
 		}
 	}
 
-	options, err := auth.getterOptions()
+	regClient, options, err := auth.registryClientAndGetterOptions()
 	if err != nil {
 		return "", err
 	}
@@ -183,6 +183,7 @@ func (h HelmClient) DownloadChart(url string, chartName string, version string, 
 		RepositoryConfig: h.settings.RepositoryConfig,
 		RepositoryCache:  h.settings.RepositoryCache,
 		Getters:          h.getterProviders,
+		RegistryClient:   regClient,
 		Options:          options,
 	}
 

--- a/pkg/applications/helmclient/client_test.go
+++ b/pkg/applications/helmclient/client_test.go
@@ -75,7 +75,10 @@ func TestDownloadChart(t *testing.T) {
 	chartArchiveDir := t.TempDir()
 
 	chartGlobPath := path.Join(chartArchiveDir, "*.tgz")
-	chartArchiveSize := packageChart(t, "testdata/examplechart", chartArchiveDir)
+	chartArchiveV1Size := packageChart(t, "testdata/examplechart", chartArchiveDir)
+	chartArchiveV2Size := packageChart(t, "testdata/examplechart-v2", chartArchiveDir)
+	chartArchiveV1Name := "examplechart-0.1.0.tgz"
+	chartArchiveV2Name := "examplechart-0.2.0.tgz"
 
 	srv, err := repotest.NewTempServerWithCleanup(t, chartGlobPath)
 	if err != nil {
@@ -96,31 +99,44 @@ func TestDownloadChart(t *testing.T) {
 	ociregistryWithAuthUrl, registryConfigFile := startOciRegistryWithAuth(t, chartGlobPath)
 
 	testCases := []struct {
-		name              string
-		repoUrl           string
-		chartName         string
-		chartVersion      string
-		auth              AuthSettings
-		expectedChartSize int64
-		wantErr           bool
+		name               string
+		repoUrl            string
+		chartName          string
+		chartVersion       string
+		auth               AuthSettings
+		expectedAchiveName string
+		expectedChartSize  int64
+		wantErr            bool
 	}{
 		{
-			name:              "Download from HTTP repository should be successful",
-			repoUrl:           srv.URL(),
-			chartName:         "examplechart",
-			chartVersion:      "0.1.0",
-			auth:              AuthSettings{},
-			expectedChartSize: chartArchiveSize,
-			wantErr:           false,
+			name:               "Download from HTTP repository should be successful",
+			repoUrl:            srv.URL(),
+			chartName:          "examplechart",
+			chartVersion:       "0.1.0",
+			auth:               AuthSettings{},
+			expectedAchiveName: chartArchiveV1Name,
+			expectedChartSize:  chartArchiveV1Size,
+			wantErr:            false,
 		},
 		{
-			name:              "Download from HTTP repository with auth should be successful",
-			repoUrl:           srvWithAuth.URL(),
-			chartName:         "examplechart",
-			chartVersion:      "0.1.0",
-			auth:              AuthSettings{Username: "username", Password: "password"},
-			expectedChartSize: chartArchiveSize,
-			wantErr:           false,
+			name:               "Download from HTTP repository with empty version should get the latest version",
+			repoUrl:            srv.URL(),
+			chartName:          "examplechart",
+			chartVersion:       "",
+			auth:               AuthSettings{},
+			expectedAchiveName: chartArchiveV2Name,
+			expectedChartSize:  chartArchiveV2Size,
+			wantErr:            false,
+		},
+		{
+			name:               "Download from HTTP repository with auth should be successful",
+			repoUrl:            srvWithAuth.URL(),
+			chartName:          "examplechart",
+			chartVersion:       "0.1.0",
+			auth:               AuthSettings{Username: "username", Password: "password"},
+			expectedAchiveName: chartArchiveV1Name,
+			expectedChartSize:  chartArchiveV1Size,
+			wantErr:            false,
 		},
 		{
 			name:              "Download from HTTP repository should fail when chart does not exist",
@@ -131,30 +147,60 @@ func TestDownloadChart(t *testing.T) {
 			expectedChartSize: 0,
 			wantErr:           true,
 		},
+		{
+			name:              "Download from HTTP repository should fail when version is not a semversion",
+			repoUrl:           srv.URL(),
+			chartName:         "examplechart",
+			chartVersion:      "notSemver",
+			auth:              AuthSettings{},
+			expectedChartSize: 0,
+			wantErr:           true,
+		},
 
 		{
-			name:              "Download from OCI repository should be successful",
-			repoUrl:           ociRegistryUrl,
-			chartName:         "examplechart",
-			chartVersion:      "0.1.0",
-			auth:              AuthSettings{},
-			expectedChartSize: chartArchiveSize,
-			wantErr:           false,
+			name:               "Download from OCI repository should be successful",
+			repoUrl:            ociRegistryUrl,
+			chartName:          "examplechart",
+			chartVersion:       "0.1.0",
+			auth:               AuthSettings{},
+			expectedAchiveName: chartArchiveV1Name,
+			expectedChartSize:  chartArchiveV1Size,
+			wantErr:            false,
 		},
 		{
-			name:              "Download from OCI repository with auth should be successful",
-			repoUrl:           ociregistryWithAuthUrl,
-			chartName:         "examplechart",
-			chartVersion:      "0.1.0",
-			auth:              AuthSettings{RegistryConfigFile: registryConfigFile},
-			expectedChartSize: chartArchiveSize,
-			wantErr:           false,
+			name:               "Download from oci repository with empty version should get the latest version",
+			repoUrl:            ociRegistryUrl,
+			chartName:          "examplechart",
+			chartVersion:       "",
+			auth:               AuthSettings{},
+			expectedAchiveName: chartArchiveV2Name,
+			expectedChartSize:  chartArchiveV2Size,
+			wantErr:            false,
+		},
+		{
+			name:               "Download from OCI repository with auth should be successful",
+			repoUrl:            ociregistryWithAuthUrl,
+			chartName:          "examplechart",
+			chartVersion:       "0.1.0",
+			auth:               AuthSettings{RegistryConfigFile: registryConfigFile},
+			expectedAchiveName: chartArchiveV1Name,
+			expectedChartSize:  chartArchiveV1Size,
+			wantErr:            false,
 		},
 		{
 			name:              "Download from oci repository should fail when chart does not exist",
 			repoUrl:           ociRegistryUrl,
 			chartName:         "chartthatdoesnotexist",
 			chartVersion:      "0.1.0",
+			auth:              AuthSettings{},
+			expectedChartSize: 0,
+			wantErr:           true,
+		},
+		{
+			name:              "Download from oci repository should fail when version is not a semversion",
+			repoUrl:           ociRegistryUrl,
+			chartName:         "examplechart",
+			chartVersion:      "notSemver",
 			auth:              AuthSettings{},
 			expectedChartSize: 0,
 			wantErr:           true,
@@ -186,8 +232,7 @@ func TestDownloadChart(t *testing.T) {
 				}
 
 				// Test chart is downloaded where we expect
-				chartArchiveName := tc.chartName + "-" + tc.chartVersion + ".tgz"
-				expectedChartLoc := downloadDest + "/" + chartArchiveName
+				expectedChartLoc := downloadDest + "/" + tc.expectedAchiveName
 				if chartLoc != expectedChartLoc {
 					t.Fatalf("charLoc is invalid. got '%s'. expect '%s'", chartLoc, expectedChartLoc)
 				}
@@ -332,6 +377,34 @@ func TestBuildDependencies(t *testing.T) {
 			hasLockFile:  false,
 			auth:         AuthSettings{RegistryConfigFile: registryConfigFile},
 			wantErr:      false,
+		},
+		{
+			name:         "http dependency with empty version should fail",
+			dependencies: []*chart.Dependency{{Name: "examplechart", Version: "", Repository: srv.URL()}},
+			hasLockFile:  false,
+			auth:         AuthSettings{},
+			wantErr:      true,
+		},
+		{
+			name:         "oci dependency with empty version should fail",
+			dependencies: []*chart.Dependency{{Name: "examplechart", Version: "", Repository: ociRegistryWithAuthUrl}},
+			hasLockFile:  false,
+			auth:         AuthSettings{},
+			wantErr:      true,
+		},
+		{
+			name:         "http dependency with non semver should fail",
+			dependencies: []*chart.Dependency{{Name: "examplechart", Version: "", Repository: srv.URL()}},
+			hasLockFile:  false,
+			auth:         AuthSettings{},
+			wantErr:      true,
+		},
+		{
+			name:         "oci dependency with non semver should fail",
+			dependencies: []*chart.Dependency{{Name: "examplechart", Version: "", Repository: ociRegistryWithAuthUrl}},
+			hasLockFile:  false,
+			auth:         AuthSettings{},
+			wantErr:      true,
 		},
 	}
 	for _, tc := range testCases {

--- a/pkg/applications/helmclient/testdata/examplechart-v2/.helmignore
+++ b/pkg/applications/helmclient/testdata/examplechart-v2/.helmignore
@@ -1,0 +1,21 @@
+# Patterns to ignore when building packages.
+# This supports shell glob matching, relative path matching, and
+# negation (prefixed with !). Only one pattern per line.
+.DS_Store
+# Common VCS dirs
+.git/
+.gitignore
+.bzr/
+.bzrignore
+.hg/
+.hgignore
+.svn/
+# Common backup files
+*.swp
+*.bak
+*.tmp
+*~
+# Various IDEs
+.project
+.idea/
+*.tmproj

--- a/pkg/applications/helmclient/testdata/examplechart-v2/Chart.yaml
+++ b/pkg/applications/helmclient/testdata/examplechart-v2/Chart.yaml
@@ -1,0 +1,17 @@
+# Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+apiVersion: v1
+description: A Helm chart for Kubernetes
+name: examplechart
+version: 0.2.0

--- a/pkg/applications/helmclient/testdata/examplechart-v2/templates/configmap.yaml
+++ b/pkg/applications/helmclient/testdata/examplechart-v2/templates/configmap.yaml
@@ -1,0 +1,26 @@
+# Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: testcm
+  labels:
+{{ if .Values.versionLabel }}
+    versionLabel: {{ .Values.versionLabel | quote}}
+{{ end }}
+data:
+{{ if .Values.cmData }}
+{{ toYaml .Values.cmData | nindent 2 }}
+{{ end }}

--- a/pkg/applications/helmclient/testdata/examplechart-v2/values.yaml
+++ b/pkg/applications/helmclient/testdata/examplechart-v2/values.yaml
@@ -1,0 +1,23 @@
+# Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Default values for examplechart.
+# This is a YAML-formatted file.
+# Declare name/value pairs to be passed into your templates.
+# name: value
+
+cmData:
+  foo: bar
+
+versionLabel: "1.0"

--- a/pkg/applications/providers/source/helm.go
+++ b/pkg/applications/providers/source/helm.go
@@ -54,7 +54,7 @@ func (h HelmSource) DownloadSource(destination string) (string, error) {
 	}
 	defer util.CleanUpHelmTempDir(helmCacheDir, h.Log)
 
-	auth, err := util.AuthFromCredentials(h.Ctx, h.SeedClient, path.Join(helmCacheDir, "reg-creg"), h.SecretNamespace, h.Source)
+	auth, err := util.HelmAuthFromCredentials(h.Ctx, h.SeedClient, path.Join(helmCacheDir, "reg-creg"), h.SecretNamespace, h.Source)
 	if err != nil {
 		return "", err
 	}

--- a/pkg/applications/providers/template/helm.go
+++ b/pkg/applications/providers/template/helm.go
@@ -63,7 +63,7 @@ func (h HelmTemplate) InstallOrUpgrade(chartLoc string, applicationInstallation 
 	}
 	defer util.CleanUpHelmTempDir(helmCacheDir, h.Log)
 
-	auth, err := util.AuthFromCredentials(h.Ctx, h.SeedClient, path.Join(helmCacheDir, "reg-creg"), h.SecretNamespace, h.ApplicationInstallation.Status.ApplicationVersion.Template.Source.Helm)
+	auth, err := util.HelmAuthFromCredentials(h.Ctx, h.SeedClient, path.Join(helmCacheDir, "reg-creg"), h.SecretNamespace, h.ApplicationInstallation.Status.ApplicationVersion.Template.Source.Helm)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/applications/providers/util/util.go
+++ b/pkg/applications/providers/util/util.go
@@ -65,8 +65,8 @@ func GetCredentialFromSecret(ctx context.Context, client ctrlruntimeclient.Clien
 	return string(cred), nil
 }
 
-// AuthFromCredentials builds helmclient.AuthSettings from source.Credentials. registryConfigFilePath is the path of the file that stores credentials for OCI registry.
-func AuthFromCredentials(ctx context.Context, client ctrlruntimeclient.Client, registryConfigFilePath string, secretNamespace string, source *appskubermaticv1.HelmSource) (helmclient.AuthSettings, error) {
+// HelmAuthFromCredentials builds helmclient.AuthSettings from source.Credentials. registryConfigFilePath is the path of the file that stores credentials for OCI registry.
+func HelmAuthFromCredentials(ctx context.Context, client ctrlruntimeclient.Client, registryConfigFilePath string, secretNamespace string, source *appskubermaticv1.HelmSource) (helmclient.AuthSettings, error) {
 	auth := helmclient.AuthSettings{}
 	if source.Credentials != nil {
 		if source.Credentials.Username != nil {

